messaging/app/src/main/java/com/google/firebase/quickstart/fcm/kotlin/ComposeMainActivity.kt

```diff
@@ -0,0 +1,267 @@
+import com.google.firebase.quickstart.fcm.R
+import com.google.firebase.quickstart.fcm.kotlin.data.SubscriptionState
+import com.google.firebase.quickstart.fcm.kotlin.ui.theme.FirebaseMessagingTheme
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+class ComposeMainActivity : ComponentActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContent {
+            FirebaseMessagingTheme {
+                // A surface container using the 'background' color from the theme
+                Surface(
+                    modifier = Modifier.fillMaxSize(),
+                    color = MaterialTheme.colors.background
+                ) {
+                    MainAppView()
+                }
+            }
+        }
+    }
+}
+
+
+@OptIn(ExperimentalPermissionsApi::class)
+@RequiresApi(Build.VERSION_CODES.TIRAMISU)
+@Composable
+private fun RequestNotificationPermissionDialog(
+    scope: CoroutineScope,
+    snackbarHostState: SnackbarHostState
+) {
+    val permissionState = rememberPermissionState(permission = Manifest.permission.POST_NOTIFICATIONS)
+
+    val requestPermissionLauncher = rememberLauncherForActivityResult(
+        ActivityResultContracts.RequestPermission()
+    ) { isGranted ->
+        var message = "Notifications permission granted"
+        if (!isGranted) {
+            message = "FCM can't post notifications without POST_NOTIFICATIONS permission"
+        }
+
+        scope.launch {
+            snackbarHostState.showSnackbar(message)
+        }
+    }
+
+    LaunchedEffect(permissionState) {
+        if (!permissionState.status.isGranted) {
+            requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
+        }
+    }
+}
+
+@Composable
+fun MainAppView(
+    lifecycleOwner: LifecycleOwner = LocalLifecycleOwner.current,
+    fcmViewModel: FirebaseMessagingViewModel = viewModel(factory = FirebaseMessagingViewModel.Factory)
+) {
+    val context = LocalContext.current
+    val scope = rememberCoroutineScope()
+    val snackbarHostState = remember { SnackbarHostState() }
+
+    val activity = context.findActivity()
+    val intent = activity?.intent
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        RequestNotificationPermissionDialog(scope, snackbarHostState)
+    }
+
+    DisposableEffect(lifecycleOwner) {
+        val observer = LifecycleEventObserver { _, event ->
+            if (event == Lifecycle.Event.ON_CREATE) {
+                //Create Notification Channel
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    fcmViewModel.setNotificationChannel(
+                        context,
+                        context.getString(R.string.default_notification_channel_id),
+                        context.getString(R.string.default_notification_channel_name)
+                    )
+                }
+
+                if (intent != null) {
```

Comment:
nit: we can use the more idiomatic Kotlin `.let` function:
```suggestion
                intent?.let {
```